Update and standardize implementation of RADIUSS packages

adrienbernede opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

This PR :
- migrates CARE and Caliper to CachedCMakePackage, closing the gap with implementations found in llnl/radiuss-spack-configs.
- adds latest releases.
- improves coherency in version constraints across RADIUSS packages.
- adds chai and care in radiuss stack for CI
- adds myself as maintainer in modified packages, fixes license when necessary, and other small changes.

adrienbernede committed:
Add latest releases of Camp, RAJA, Umpire, CHAI and CARE

adrienbernede commented:
Forced push to rebase on latest snapshot release so that CI can run.

adayton1 commented on var/spack/repos/builtin/packages/umpire/package.py:
```diff
@@ -22,8 +22,13 @@ class Umpire(CachedCMakePackage, CudaPackage, ROCmPackage):
     maintainers("davidbeckingsale", "adrienbernede")
 
     license("MIT")
-
```
Keep this extra line between the license and all the versions.

adrienbernede replied:
Done

adayton1 reviewed (approved):
LGTM

adayton1 commented on var/spack/repos/builtin/packages/care/package.py:
```diff
@@ -21,6 +21,18 @@ class Care(CMakePackage, CudaPackage, ROCmPackage):
+        "0.13.3",
+        tag="v0.13.3",
+        commit="93853696b452647278eae9311b835ad206236522",
+        submodules="True",
```
I don't think I understood this before, but I don't think we need the submodules.

adrienbernede replied:
OK, will have to test that change in our CI.
Would you know whether they were needed at some point ?

adayton1 replied:
I'm not sure they were ever needed.

adrienbernede committed:
Address review comments + blt requirement in Umpire

adrienbernede commented:
@adayton1 would you know why CHAI is not finding RAJA anymore ?
https://gitlab.spack.io/spack/spack/-/jobs/12157797

adayton1 commented:
> @adayton1 would you know why CHAI is not finding RAJA anymore ? https://gitlab.spack.io/spack/spack/-/jobs/12157797

It looks like RAJA was never installed in the job. So it seems that it recognized that the RAJA variant was requested, but never picked up the dependency. Maybe there needs to be a simple depends_on("raja") line?

v-dobrev reviewed (dismissed):
The changes to the `mfem` package look good to me.

adrienbernede committed:
CARE @develop & @main: Submodules -> False

adrienbernede committed:
Merge tag 'develop-2024-09-01' into woptim/radiuss-packages-update

Snapshot release 2024-09-01

adrienbernede committed:
Changes in Umpire

adrienbernede committed:
Changes in RAJA

adrienbernede committed:
Changes in CHAI

adrienbernede committed:
Changes in RAJA: prefer 'spec.satisfies' to 'in spec'

This is due to a non-equivalence in Spack with providers like mpi.
See e.g. https://github.com/spack/spack/pull/46126

adrienbernede committed:
Changes in Umpire: prefer 'spec.satisfies' to 'in spec'

This is due to a non-equivalence in Spack with providers like mpi.
See e.g. https://github.com/spack/spack/pull/46126

adrienbernede committed:
Changes in CARE:
Still need to update to CachedCMakePackage based on RADIUSS Spack Configs version

adrienbernede committed:
Missing change in RAJA + changes in fmt

adrienbernede committed:
Fix syntax

adrienbernede committed:
Changes in Camp

adrienbernede committed:
Fix style

adrienbernede commented:
@adayton1 The failure with CHAI is due to CHAI build system still looking for RAJA while the variant is set to `~raja`...

adrienbernede committed:
CHAI: when ~raja, turn off RAJA in build system

adrienbernede committed:
Fix: Ascent@0.9.3 does not support RAJA@2024.07.0

adrienbernede commented:
@cyrush I’d like your validation regarding the requirement on the RAJA version in Ascent.

cyrush commented:
@adrienbernede  here are the details for Ascent 0.9.3:

*conduit@0.9.1 (newer should be fine)
* vtk-m@2.1.0 (requires [patch](https://github.com/Alpine-DAV/ascent/blob/0aef6cffd522be7419651e6adf586f9a553297d0/scripts/build_ascent/2024_05_03_vtkm-mr3215-ext-geom-fix.patch) )
* raja@2024.02.1
* umpire@2024.02.1
* camp@2024.02.1
* kokkos@3.7.02

These are from:

https://github.com/Alpine-DAV/ascent/blob/develop/CHANGELOG.md#093---released-2024-05-11

adayton1 commented on var/spack/repos/builtin/packages/chai/package.py:
```diff
@@ -259,35 +270,43 @@ def initconfig_package_entries(self):
             entries.append(cmake_cache_option("{}ENABLE_RAJA_PLUGIN".format(option_prefix), True))
             entries.append(cmake_cache_path("RAJA_DIR", spec["raja"].prefix))
+        else:
+            entries.append(cmake_cache_option("{}ENABLE_RAJA_PLUGIN".format(option_prefix), False))
```
Did this fix the CHAI failures you were seeing?

adrienbernede replied:
Yes

adrienbernede commented:
@cyrush you are listing the "Preferred dependency versions for ascent@0.9.3".
Likely, it means that anything newer is not guaranteed to be supported.
In which case I should apply to Umpire and Campe the same version constraint I added for RAJA.
Is that what you are suggesting ?

adrienbernede committed:
Enforce same version constraint on Umpire as for RAJA

adrienbernede committed:
Merge tag 'develop-2024-09-08' into woptim/radiuss-packages-update

Snapshot release 2024-09-08

cyrush commented:
@adrienbernede  That shows what we have tested & we can't really guarantee Ascent will work with over versions.  If we had testing in spack we could try to loosen, but I don't think we have good testing to try an experiment with any version.

adrienbernede commented:
@cyrush Understood. Then I think I should also have Ascent@0.9.3 explicitly require vtk-m@2.1.0, which is not the case currently.

adrienbernede committed:
Enforce preferred version of vtk-m in ascent 0.9.3

adrienbernede commented:
Note: ascent being built with vtk-m 2.2 in Spack CI, the commit above had no effect (it's only requiring vtk-m > 2.1).

@cyrush per your comment, I’m tempted to restrict vtk-m to 2.1.0:2.1.99. It builds with 2.2, but hasn't been tested.

adrienbernede committed:
Migrate CARE package to CachedCMakePackage

adrienbernede committed:
Fix style in CARE package

adrienbernede committed:
CARE: Apply changes for uniform implementation accross RADIUSS projects

adrienbernede committed:
Caliper: move to CachedCMakePackage, from RADIUSS Spack Configs

adrienbernede committed:
Adapt RAJA Perf to spack CI

adrienbernede committed:
Activate CHAI, CARE and RAJAPerf in Spack CI

adrienbernede committed:
Fixes and diffs with RADIUSS Spack Configs

adrienbernede committed:
Caliper: fix

adrienbernede committed:
Caliper: fix

adrienbernede committed:
Caliper : fix + RAJAPerf : style

adrienbernede committed:
RAJAPerf: fixes

adrienbernede committed:
Update maintainers

adrienbernede committed:
raja-perf: Fix license header

adrienbernede committed:
raja-perf: Fix variant naming openmp_target -> omptarget

adrienbernede committed:
raja-perf: style and blt dependency versions

adrienbernede committed:
CARE: benchmark and examples off by default (like tests)

adrienbernede committed:
CARE: fix missing variable

adrienbernede commented:
@tldahlgren This PR is ready, I’d like you review on it (of another Spack team member).

adrienbernede commented:
@adayton1 please note the migration of CARE package to CachedCMakePackage (based on implementation in radiuss-spack-configs)
@daboehme please note the migration of Caliper package to CachedCMakePackage (based on implementation in radiuss-spack-configs)
@rhornung67 please note the addition of RAJAPerf package (based on implementation in radiuss-spack-conifgs)

tldahlgren commented on var/spack/repos/builtin/packages/raja-perf/package.py:
```diff
@@ -0,0 +1,325 @@
+class RajaPerf(CachedCMakePackage, CudaPackage, ROCmPackage):
+    """RAJA Performance Suite."""
+
+    homepage = "http://software.llnl.gov/RAJAPerf/"
```
This is no longer a valid URL.

tldahlgren replied:
The "closest" non-github URL may be https://software.llnl.gov/#RAJAPerf BUT you still have to search for `RAJAPerf` to narrow it down to the software.

tldahlgren replied:
It isn't available from https://software.llnl.gov/radiuss/projects/, only `raja` yields the software "box" link.

adrienbernede replied:
Done !

tldahlgren commented on var/spack/repos/builtin/packages/care/package.py:
```diff
@@ -3,42 +3,59 @@
-    version("develop", branch="develop", submodules="True")
-    version("master", branch="main", submodules="True")
+    version("develop", branch="develop", submodules=False)
+    version("master", branch="main", submodules=False)
```
It looks like the branch is actually `master`, not `main` (and there is no `main` branch ATM).

adrienbernede replied:
Done

tldahlgren reviewed (changes requested):
Confirmed all affected version commits (or tags) *except* `care@master` (see comment).  There's also an issue with the URL for `raja-perf`.

FWIW. Some people are hesitant to change variant defaults since others *may* depend on them.

rhornung67 commented on var/spack/repos/builtin/packages/raja-perf/package.py:
```diff
@@ -0,0 +1,325 @@
+class RajaPerf(CachedCMakePackage, CudaPackage, ROCmPackage):
+    """RAJA Performance Suite."""
+
+    homepage = "http://software.llnl.gov/RAJAPerf/"
```
```suggestion
    homepage = "https://github.com/LLNL/RAJAPerf"
```

adrienbernede committed:
Update var/spack/repos/builtin/packages/raja-perf/package.py

Co-authored-by: Rich Hornung <hornung1@llnl.gov>

adrienbernede committed:
CARE: fix branch name

tldahlgren reviewed (dismissed):
Affected homepages, branches, commits/tags LGTM.

adrienbernede commented on var/spack/repos/builtin/packages/care/package.py:
```diff
@@ -50,137 +67,219 @@ class Care(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("cxx", type="build")  # generated
     depends_on("fortran", type="build")  # generated
 
-    variant("openmp", default=False, description="Build Shared Libs")
+    variant("openmp", default=False, description="Build with OpenMP support")
+    variant("mpi", default=False, description="Enable MPI support")
     variant(
         "implicit_conversions",
-        default=True,
+        default=False,
```
@adayton1 I didn't even notice this change I made. We had implicit_conversions set to "False" locally. Would you prefer to have it to "True" here in Spack ?
Thanks @tldahlgren for the warning.

adayton1 replied:
I think it should default to False.

adrienbernede commented on var/spack/repos/builtin/packages/care/package.py:
```diff
@@ -50,137 +67,219 @@ class Care(CMakePackage, CudaPackage, ROCmPackage):
-    variant("examples", default=True, description="Build examples.")
-    variant("docs", default=False, description="Build documentation")
     variant("tests", default=False, description="Build tests")
+    variant("benchmarks", default=False, description="Build benchmarks.")
+    variant("examples", default=False, description="Build examples.")
```
I set those to "False" which is coherent with `tests` being set to "False" itself.
Especially true for `benchmarks` which requires `tests`.
@adayton1 Are you OK with that ?

adrienbernede replied:
Please also validate "docs" below... maybe I shouldn't have changed that one...

adayton1 replied:
I'm OK with tests, benchmarks, examples, and docs all defaulting to False.

adayton1 reviewed (approved):
LGTM

tldahlgren commented:
Is this PR ready to merge assuming all of the checks pass?

adrienbernede committed:
Merge tag 'develop-2024-09-22' into woptim/radiuss-packages-update

Snapshot release 2024-09-22

adrienbernede commented:
@tldahlgren @v-dobrev I don't think the error with MFEM is caused by this pull request since it does not affect the MFEM package features nor any of MFEM dependencies involved in the build.
Instead, I think the MFEM build was triggered because the package was touched, but the error was introduced by another PR which hadn't triggered the MFEM builds.
```
  >> 590    ld.lld: error: undefined reference due to --no-allow-shlib-undefine
            d: dlclose
     591    >>> referenced by /opt/cray/pe/libsci/23.02.1.1/CRAY/9.0/x86_64/lib
            /libsci_cray.so
```

adrienbernede committed:
Revert changes in MFEM to pass CI

adrienbernede commented:
> Is this PR ready to merge assuming all of the checks pass?

Yes it is.

adrienbernede commented:
@tldahlgren @v-dobrev I reverted the changes in the MFEM package and applied them in a separate PR.

https://github.com/spack/spack/pull/46537

CI is now passing.

adrienbernede commented:
@tldahlgren this is ready (only failure is coverage) ;)

adrienbernede committed:
Fix CXX17 condition in RAJA + add sycl option in RAJAPerf

adrienbernede commented:
@trws, @tldahlgren being out of office this week, would you mind helping me getting this merged ?

trws commented:
I'm just getting back from Australia so I'm not sure I'll be faster, but I can try.  What do you still need?

adrienbernede commented:
@trws I need you to hit the "approve" button. ;)
